Tests/FluentAssertions.Specs/Collections/Data/DataColumnCollectionAssertionExtensionsSpecs.cs

```diff
@@ -0,0 +1,420 @@
+
+                // Assert
+                action.Should().Throw<XunitException>().WithMessage(
+                    "Expected genericCollection to refer to DataColumnCollection because we care, but found * (different type).");
+            }
+        }
+
+        public class NotBeSameAs
+        {
+            [Fact]
+            public void When_references_are_the_same_it_should_fail()
+            {
+                // Arrange
+                var dataTable = new DataTable("Test");
+
+                var columnCollection1 = dataTable.Columns;
+                var columnCollection2 = columnCollection1;
+
+                // Act
+                Action action =
+                    () => columnCollection1.Should().NotBeSameAs(columnCollection2);
+
+                // Assert
+                action.Should().Throw<XunitException>().WithMessage("Did not expect columnCollection1 to refer to *.");
+            }
+
+            [Fact]
+            public void When_references_are_different_it_should_succeed()
+            {
+                // Arrange
+                var dataTable1 = new DataTable("Test1");
+                var dataTable2 = new DataTable("Test2");
+
+                var columnCollection1 = dataTable1.Columns;
+                var columnCollection2 = dataTable2.Columns;
+
+                // Act & Assert
+                columnCollection1.Should().NotBeSameAs(columnCollection2);
+            }
+        }
+
+        public class HaveSameCount
+        {
+            [Fact]
+            public void When_subject_is_null_it_should_fail()
+            {
+                // Arrange
+                var subject = default(DataColumnCollection);
+
+                var expectation = new DataTable().Columns;
+
+                // Act
+                Action action =
+                    () => subject.Should().HaveSameCount(expectation, because: "we {0}", "care");
+
+                // Assert
+                action.Should().Throw<XunitException>().WithMessage(
+                    "Expected * to have the same count as * because we care, but found <null>.*");
+            }
+
+            [Fact]
+            public void When_expectation_is_null_it_should_fail()
+            {
+                // Arrange
+                var dataTable = new DataTable();
+
+                dataTable.Columns.Add(new DataColumn("Column0"));
+                dataTable.Columns.Add(new DataColumn("Column1"));
+                dataTable.Columns.Add(new DataColumn("Column2"));
+
+                var nullReference = default(DataColumnCollection);
+
+                // Act
+                Action action =
+                    () => dataTable.Columns.Should().HaveSameCount(nullReference);
+
+                // Assert
+                action.Should().Throw<ArgumentNullException>().WithMessage(
+                    "Cannot verify count against a <null> collection.*");
+            }
+
+            public class DataColumnCollectionAssertions
```

Comment:
🤔 Why do you need `DataColumnCollectionAssertions` and `GeneicCollectionAssertions`?

Reply:
When they were rwitten, they are testing different methods in the subject. As that's no longer the case, I'll simplify this.

Reply:
Oh, actually, even though these are calling the same method, the `.Should()` call follows a different path and produces a different type of subject object. In one set of tests, the subject is a `ReadOnlyNonGenericCollectionWrapper<TCollection, TItem>`, while the in the other, the subject is simply an `IEnumerable<T>`.

Reply:
Then I would keep the nested class at the same level as the others.

Reply:
No, on second thought. You're still testing the `HaveSameCount` API, so it belongs in the same nested class that is used for grouping per API. The fact that the subject doesn't matter. That's just one of the many test scenarios for that API.

Reply:
So, to sketch out what you're proposing, instead of this:
```
public static class DataColumnCollectionAssertionSpecs
{
  public class HaveSameCount
  {
    public class DataColumnCollectionAssertions
    {
      [Fact]
      public void Test() { }
    }

    public class GenericCollectionAssertions
    {
      [Fact]
      public void Test() { }
    }
  }
}
```
...it'd look like this?:
```
public static class DataColumnCollectionAssertionSpecs
{
  public class HaveSameCount
  {
    [Fact]
    public void Test_with_DataColumnCollection() { }
 
    [Fact]
    public void Test_with_generic_collection() { }
  }
}
```
How to work around the desire to avoid putting the type name into the test name??

Reply:
You can't really avoid that. But I would write it as `data_column_collection`.

Reply:
Doesn't that just obfuscate the fact that it is, in fact, referring to a specific data type??

Reply:
Yes, but that's clear from the implementation of the test. "data column collection" is a pretty generic name.

Reply:
So you'd like me to take the tests that are in parallel depending on the compile-time type of the subject, and currently separated out into nested classes, and simply drop the nested classes so that they're just bunched together into the parent type, and then instead of using the nested classes, put an indicator of the subject type into the test name?? I can do that but it feels like a step backward in organization & readability.

Reply:
This is now implemented in this way, and the latest two commits, 0c7c715 and 0216b32, show the conversion back to an assertion instead of a direct `throw` with a reusable `FormatFailureMessage` in `AssertionScope`. The code is very parallel; the change is pretty small.

Reply:
I'm not sure I understand what you're saying, but in general, we stick to one level of nested classes that are either named to the API (e.g. `HaveSameCount`) or use a more functional name (`MemberMapping` in the `BeEquivalentTo` specs).

Reply:
Well, in this case, the methods in question have different paths for different types of subject, and these paths need essentially the same tests. If I limit it to one level of nested classes, then those different paths get dropped into the same basket. They need very long names to spell out what exactly is being tested.

I'm not sure what you mean by `MemberMapping` -- I can't find that anywhere in the codebase. But, in this instance, there's essentially a cartesian relationship between the tests and the types. One way or the other, they need to extend across two axes, and some organization reflecting this significantly improves the readability of the code, in my opinion.

Reply:
Yeah, I was wrong with the `MemberMapping` example. Those are even in a different file called `MemberMappingSpecs`. You know what, it's not important enough, so stick to the nested classes you had in mind originally.

Reply:
I can't actually find `MemberMapping` anywhere in the source tree, for what it's worth. Was it added very recently, perhaps, since the last rebase??

Reply:
I meant that it wasn't a nested class. It became an entirely separate test class called `MemberMappingSpecs`.

Reply:
Yeah, but...
```
[C:\code\fluentassertions-2]ffind /S/T"MemberMapping"/V *.cs
0 lines in 0 files

[C:\code\fluentassertions-2]
```